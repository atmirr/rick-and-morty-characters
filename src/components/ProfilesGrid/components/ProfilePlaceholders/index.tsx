import React from 'react';
import styled from 'styled-components';
import ContentLoader from 'react-content-loader';

const PlaceholderWrapper = styled.div`
  margin: 1.1em 0.7em;
  border: 3px solid #f3f3f3;
  border-radius: 15px;
`;

type PropTypes = {
  count: number;
  [key: string]: string | number;
};

function ProfilePlaceholders({
  count = 1,
  ...props
}: PropTypes): React.ReactElement {
  const ProfilePlaceholder = (props: { [key: string]: string | number }) => (
    <PlaceholderWrapper>
      <ContentLoader
        speed={1}
        width={273}
        height={500}
        viewBox="0 0 273 500"
        backgroundColor="#f3f3f3"
        foregroundColor="#ecebeb"
        {...props}
      >
        <rect x="15" y="15" rx="15" ry="15" width="243" height="200" />
        <rect x="15" y="239" rx="0" ry="0" width="243" height="10" />
        <rect x="15" y="264" rx="0" ry="0" width="243" height="10" />
        <rect x="15" y="289" rx="0" ry="0" width="243" height="10" />
        <rect x="15" y="314" rx="0" ry="0" width="243" height="10" />
        <rect x="15" y="339" rx="0" ry="0" width="243" height="10" />
        <rect x="15" y="364" rx="0" ry="0" width="243" height="10" />
        <rect x="15" y="389" rx="0" ry="0" width="243" height="10" />
        <rect x="15" y="414" rx="0" ry="0" width="243" height="10" />
        <rect x="15" y="439" rx="0" ry="0" width="243" height="10" />
        <rect x="15" y="464" rx="0" ry="0" width="243" height="10" />
      </ContentLoader>
    </PlaceholderWrapper>
  );
  const PlaceholderItems = [];
  for (let i = 0; i < count; i++) {
    PlaceholderItems.push(
      <ProfilePlaceholder
        key={i}
        uniqueKey={`profile-placeholder-${i}`}
        data-testid={`profile-placeholder-${i}`}
        {...props}
      />,
    );
  }
  return <>{PlaceholderItems}</>;
}

export default ProfilePlaceholders;
