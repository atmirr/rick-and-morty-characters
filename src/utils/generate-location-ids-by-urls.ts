export const generateLocationIdsByUrls = (urls: string[]): string[] =>
  urls.reduce((result: string[], url: string) => {
    const splittedUrl = url.split('/');
    const id = splittedUrl[splittedUrl.length - 1];
    return id ? [...result, id] : [...result];
  }, []);
