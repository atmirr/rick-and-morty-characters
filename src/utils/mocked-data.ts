export const characters = [
  {
    id: 1,
    name: 'Rick Sanchez',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:48:46.250Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/1',
  },
  {
    id: 2,
    name: 'Morty Smith',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:50:21.651Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
      'https://rickandmortyapi.com/api/episode/3',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/2',
  },
];

export const episodes = [
  'Rixty Minutes',
  'Something Ricked This Way Comes',
  'Close Rick-counters of the Rick Kind',
  'Ricksy Business',
  'A Rickle in Time',
];

export const locations = [
  {
    name: 'Citadel of Ricks',
    type: 'Space station',
    dimension: 'unknown',
  },
  {
    name: 'Earth (Replacement Dimension)',
    type: 'Planet',
    dimension: 'Replacement Dimension',
  },
];
