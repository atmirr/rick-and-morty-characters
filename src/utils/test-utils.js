import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockedInitialState = {
  characters: { items: [], nextPage: null, error: null, isLoading: false },
  episodes: { items: [], nextPage: null, error: null, isLoading: false },
  locations: { items: [], error: null, isLoading: false },
};

function render(
  ui,
  {
    initialState = mockedInitialState,
    store = configureStore()(initialState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
