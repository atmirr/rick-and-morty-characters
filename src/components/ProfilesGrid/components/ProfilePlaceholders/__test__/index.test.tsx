import React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { screen, render } from 'src/utils/test-utils';
import ProfilePlaceholders from '../index';

describe('<ProfilePlaceholders />', () => {
  it('should renders the component', () => {
    const { container } = render(<ProfilePlaceholders />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render 8 profile placeholders when the count prop has been set to 8', () => {
    render(<ProfilePlaceholders count={8} />);
    expect(
      screen.queryAllByTestId(/^(profile-placeholder-)+([0-9]{1,2})$/).length,
    ).toBe(8);
  });
});
