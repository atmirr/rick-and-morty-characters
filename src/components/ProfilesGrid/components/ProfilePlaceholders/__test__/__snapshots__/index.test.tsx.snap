// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfilePlaceholders /> should renders the component 1`] = `
.c0 {
  margin: 1.1em 0.7em;
  border: 3px solid #f3f3f3;
  border-radius: 15px;
}

<div
  class="c0"
>
  <svg
    aria-labelledby="profile-placeholder-0-aria"
    data-testid="profile-placeholder-0"
    height="500"
    role="img"
    viewBox="0 0 273 500"
    width="273"
  >
    <title
      id="profile-placeholder-0-aria"
    >
      Loading...
    </title>
    <rect
      clip-path="url(#profile-placeholder-0-diff)"
      height="100%"
      role="presentation"
      style="fill: url(#profile-placeholder-0-animated-diff);"
      width="100%"
      x="0"
      y="0"
    />
    <defs
      role="presentation"
    >
      <clippath
        id="profile-placeholder-0-diff"
      >
        <rect
          height="200"
          rx="15"
          ry="15"
          width="243"
          x="15"
          y="15"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="239"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="264"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="289"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="314"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="339"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="364"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="389"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="414"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="439"
        />
        <rect
          height="10"
          rx="0"
          ry="0"
          width="243"
          x="15"
          y="464"
        />
      </clippath>
      <lineargradient
        id="profile-placeholder-0-animated-diff"
      >
        <stop
          offset="0%"
          stop-color="#f3f3f3"
          stop-opacity="1"
        >
          <animate
            attributeName="offset"
            dur="1s"
            keyTimes="0; 0.25; 1"
            repeatCount="indefinite"
            values="-2; -2; 1"
          />
        </stop>
        <stop
          offset="50%"
          stop-color="#ecebeb"
          stop-opacity="1"
        >
          <animate
            attributeName="offset"
            dur="1s"
            keyTimes="0; 0.25; 1"
            repeatCount="indefinite"
            values="-1; -1; 2"
          />
        </stop>
        <stop
          offset="100%"
          stop-color="#f3f3f3"
          stop-opacity="1"
        >
          <animate
            attributeName="offset"
            dur="1s"
            keyTimes="0; 0.25; 1"
            repeatCount="indefinite"
            values="0; 0; 3"
          />
        </stop>
      </lineargradient>
    </defs>
  </svg>
</div>
`;
